compiler/rustc_middle/src/ty/relate.rs

```diff
@@ -54,19 +54,18 @@ impl<'tcx> Relate<TyCtxt<'tcx>> for ty::Pattern<'tcx> {
                 &ty::PatternKind::Range { start: start_a, end: end_a, include_end: inc_a },
                 &ty::PatternKind::Range { start: start_b, end: end_b, include_end: inc_b },
             ) => {
-                // FIXME(pattern_types): make equal patterns equal (`0..=` is the same as `..=`).
-                let mut relate_opt_const = |a, b| match (a, b) {
-                    (None, None) => Ok(None),
-                    (Some(a), Some(b)) => relation.relate(a, b).map(Some),
-                    // FIXME(pattern_types): report a better error
-                    _ => Err(TypeError::Mismatch),
-                };
-                let start = relate_opt_const(start_a, start_b)?;
-                let end = relate_opt_const(end_a, end_b)?;
-                if inc_a != inc_b {
-                    todo!()
+                let start = relation.relate(start_a, start_b)?;
+                // FIXME(pattern_types): make equal patterns equal (`0..5` is the same as `0..=6`).
```

Comment:
I think we'll basically want to mark all pattern types that have `..` ranges to need normalization, just like pattern types with generic consts in them? This is a problem for future me, but still sth that came up while writing the hir_ty_lowering logic